Sources/StreamChat/APIClient/APIClient.swift

```diff
@@ -90,6 +128,46 @@ class APIClient {
             completion: completion
         )
     }
+    
+    /// Queues a failed request for executing later or failing after a timeout
+    func handleTokenExpirationError<Response: Decodable>(
+        urlRequest: URLRequest,
+        timeout: TimeInterval,
+        completion: @escaping (Result<Response, Error>) -> Void
+    ) {
+        requestsAccessQueue.async {
+            let item = RequestsQueueItem(
+                requestAction: { self.request(
+                    urlRequest: urlRequest,
+                    timeout: timeout,
+                    completion: completion
+                )
+                },
+                failureAction: { completion(.failure(ClientError.ExpiredToken())) }
+            )
+            
+            // The moment we queue a first request for execution later we also set a timeout for 60 seconds
+            // after which we fail all the queued requests
+            if self.requestsQueue.isEmpty {
+                self.requestsAccessQueue.asyncAfter(deadline: .now() + timeout) {
+                    self.requestsQueue.forEach { $0.failureAction() }
+                    self.requestsQueue = []
+                }
+            }
+            
+            self.requestsQueue.append(item)
+            
+            // Only initiate token refreshing once
+            guard !self.isRefreshingToken else { return }
+            
+            self.isRefreshingToken = true
+            self.tokenRefresher(ClientError.ExpiredToken()) {
+                self.requestsQueue.forEach { $0.requestAction() }
```

Comment:
Following feedback is debatable since it only applies if the token expires quickly after one is obtained:

Order here is not 100% correct, you probably want to first swap `isRefreshingToken` to false and then consume the queue. 

It also sounds safer to consume the queue this way:

- while requestsQueue is not empty
   - check `isRefreshingToken` or exit
   - pop requestsQueue 
   - exec